### ISSUE TITLE
Most "provide" methods don't work

### ISSUE THREAD

**Issue by taion** [tavendo/AutobahnCpp#21]
The code here:

https://github.com/tavendo/AutobahnCpp/blob/b44ee2e0582923fe8e175eff1ec726582575f4e2/autobahn/autobahn_impl.hpp#L719

Only supports 3 of the implemented provide methods. It looks like the rest just don't work. Is it possible to temporarily remove the unsupported provide types for the time being?

**Comment by taion** [comment]
Also, it might make sense to have a void endpoint type for functions that don't have a return value.

**Comment by oberstet** [comment]
Only 1 is missing: return type `endpoint_v_t`.

**Comment by taion** [comment]
I don't think that's true. It looks like that check in the linked code only supports

* `endpoint_t`
* `endpoint_v_t`
* `endpoint_fvm_t`

I've checked by attempting and failing that you hit the `FIX ME INVOCATION` error with `endpoint_m_t`, but I'd guess this doesn't work with any of

* `endpoint_m_t`
* `endpoint_vm_t`
* `endpoint_f_t`
* `endpoint_fv_t`
* `endpoint_fm_t`

I can send in a PR for this if you want.

**Comment by taion** [comment]
Actually, it turns out that none of the endpoints that expect futures actually work. The issue is that you actually wait on the future here:

https://github.com/tavendo/AutobahnCpp/blob/b44ee2e0582923fe8e175eff1ec726582575f4e2/autobahn/autobahn_impl.hpp#L771

However, this ends up blocking the running thread. If my endpoint does something like

```cpp
return session.call(procedure, args).then(/* do some work*/);
```

then the procedure will never actually return, because the running thread is blocked on the `done.wait()` call, and cannot process the return value from the called procedure and resolve the future.

You can't wait for the future to be resolved there.

**Comment by oberstet** [comment]
Here is an example for `endpoint_fvm_t`: https://github.com/tavendo/AutobahnCpp/blob/master/examples/register2.cpp#L63

**Comment by taion** [comment]
The reason that example works is because the future is ready, so `done.wait()` returns immediately.

If waiting for the future actually requires handling some input (e.g. you invoke another method and wait for it to finish), then it will not work, because you've blocked the main thread from actually processing the response.

**Comment by oberstet** [comment]
I see.

I guess it would be good if we first had an example that demonstrates the issue, and then come up with a solution.

This is the harder problem - extending the solution to the variants should then be straightforward.

**Comment by taion** [comment]
The missing `provide` methods are now present, either way. I'll open a new issue regarding the problem with futures.
